File: apps/web/app/types/observability.ts
export interface EnvironmentMetrics {
  environment_id: string
  name: string
  status: string
  namespace_name: string
  cpu_cores: number
  memory_gib: number
  cpu_percent: number | null
  memory_percent: number | null
  source: string | null
  available: boolean
  detail: string | null
  sampled_at: string
}

export interface EnvironmentHealthPing {
  environment_id: string
  name: string
  status: string
  ok: boolean
  status_code: number | null
  message: string
  preview_url: string | null
  latency_ms: number | null
  checked_at: string
}

export interface EnvironmentObservabilityItem {
  environment_id: string
  name: string
  status: string
  provider: string | null
  deploy_mode: string | null
  app_ready: boolean
  preview_url: string | null
  metrics: EnvironmentMetrics
  health: EnvironmentHealthPing
}

export interface EnvironmentObservabilitySummary {
  items: EnvironmentObservabilityItem[]
  healthy_count: number
  unhealthy_count: number
  unknown_count: number
  sampled_at: string
}

File: apps/api/app/schemas/observability.py
"""Schemas for environment metrics and health-ping observability."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class EnvironmentMetricsRead(BaseModel):
    environment_id: UUID
    name: str
    status: str
    namespace_name: str
    cpu_cores: float = 0.0
    memory_gib: float = 0.0
    cpu_percent: float | None = None
    memory_percent: float | None = None
    source: str | None = None
    available: bool = False
    detail: str | None = None
    sampled_at: datetime


class EnvironmentHealthPingRead(BaseModel):
    environment_id: UUID
    name: str
    status: str
    ok: bool
    status_code: int | None = None
    message: str
    preview_url: str | None = None
    latency_ms: float | None = None
    checked_at: datetime


class EnvironmentObservabilityItem(BaseModel):
    environment_id: UUID
    name: str
    status: str
    provider: str | None = None
    deploy_mode: str | None = None
    app_ready: bool = False
    preview_url: str | None = None
    metrics: EnvironmentMetricsRead
    health: EnvironmentHealthPingRead


class EnvironmentObservabilitySummary(BaseModel):
    items: list[EnvironmentObservabilityItem] = Field(default_factory=list)
    healthy_count: int = 0
    unhealthy_count: int = 0
    unknown_count: int = 0
    sampled_at: datetime
